nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttribute.java

```diff
@@ -0,0 +1,273 @@
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@SideEffectFree
+@SupportsBatching(defaultDuration = DefaultRunDuration.TWENTY_FIVE_MILLIS)
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@Tags({"attributes", "modification", "filter", "retain", "remove", "delete", "regex", "regular expression", "Attribute Expression Language"})
+@CapabilityDescription("Filters the Attributes of a FlowFile according to a specified strategy")
+@UseCase(
+        description = "Retain all FlowFile attributes matching a regular expression",
+        configuration = """
+                Set "Filter mode" to "Retain".
+                Set "Attribute matching strategy" to "Use regular expression".
+                Specify the "Regular expression to filter attributes", e.g. "my-property|a-prefix[.].*".
+                """
+)
+@UseCase(
+        description = "Remove only a specified set of FlowFile attributes",
+        configuration = """
+                Set "Filter mode" to "Remove".
+                Set "Attribute matching strategy" to "Enumerate attributes".
+                Specify the set of "Set of attributes to filter" using the delimiter comma ',', e.g. "my-property,other,filename".
+                """
+)
+public class FilterAttribute extends AbstractProcessor {
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .description("All successful FlowFiles are routed to this relationship").name("success").build();
+
+    private final static Set<Relationship> relationships = Collections.singleton(REL_SUCCESS);
+
+
+    public static final AllowableValue FILTER_MODE_VALUE_RETAIN = new AllowableValue(
+            "RETAIN",
+            "Retain",
+            "Retains only the attributes matching the filter, all other attributes are removed."
+    );
+
+    public static final AllowableValue FILTER_MODE_VALUE_REMOVE = new AllowableValue(
+            "REMOVE",
+            "Remove",
+            "Removes the attributes matching the filter, all other attributes are retained."
+    );
+
+    public static final PropertyDescriptor FILTER_MODE = new PropertyDescriptor.Builder()
+            .name("FILTER_MODE")
+            .displayName("Filter mode")
+            .description("Specifies the strategy to apply on filtered attributes. Either 'Remove' or 'Retain' only the matching attributes.")
+            .required(true)
+            .allowableValues(FILTER_MODE_VALUE_RETAIN, FILTER_MODE_VALUE_REMOVE)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
```

Comment:
I understand the desire for FILTER_MODE to support expression language, but it presents some problems.
- the AllowableValue description can never be seen
- the operator has to type in either RETAIN or REMOVE, exactly
- if an ExpLang expression is entered, the processor is invalid because it doesn't match the allowableValues

If you remove the allowableValues restriction and replace it with a validator (see InvokeHTTP "HTTP Method"), then the processor can be valid with an ExpLang expression.  However, if that expression evaluates to an invalid FILTER_MODE, then the processor will admin yield the flowfile which stops data flowing through the processor.  You could catch this problem and send the flowfile to 'success' unmodified, but that hides the problem with the FlowFile's attributes.  You could have a failure relationship to deal with cases when the ExpLang expression causes a problem.

The easiest approach is to remove support for ExpLang for FILTER_MODE, though I don't have a guess for how many users would need this feature.

Reply:
These are some good remarks, thank you.

I assume there won't be a lot of users that wish to use EL for the FILTER_MODE. I removed the support for now. In case using EL for this property is more desired than anticipated right now, it should be relatively easy to introduce proper support in one of the ways you mentioned.